editor/plugins/polygon_2d_editor_plugin.cpp

```diff
@@ -624,16 +625,82 @@ void Polygon2DEditor::_uv_input(const Ref<InputEvent> &p_input) {
 
 				if (uv_move_current == UV_MODE_EDIT_POINT) {
 					point_drag_index = -1;
-					for (int i = 0; i < points_prev.size(); i++) {
-						Vector2 tuv = mtx.xform(points_prev[i]);
-						if (tuv.distance_to(Vector2(mb->get_position().x, mb->get_position().y)) < 8) {
-							uv_drag_from = tuv;
-							point_drag_index = i;
+					bool found_edge_pt = false;
+
+					// Check if selected point is close to an existing edge
+					const real_t grab_threshold = (8 / uv_draw_zoom);
+					const real_t eps = grab_threshold * 2;
+					const real_t eps2 = eps * eps;
+
+					Vector2 closest_pt;
+					int closest_idx = 0;
+					real_t closest_dist = 1e10;
+
+					const int n_points = points_prev.size();
+					Vector2 p_pos = mtx.affine_inverse().xform(snap_point(Vector2(mb->get_position().x, mb->get_position().y)));
+
+					for (int i = 0; i < n_points; i++) {
+						Vector2 segment[2] = { points_prev[i], points_prev[(i+1) % n_points] };
+						Vector2 cp = Geometry2D::get_closest_point_to_segment(p_pos, segment);
+						if (cp.distance_squared_to(segment[0]) < eps2 || cp.distance_squared_to(segment[1]) < eps2) {
+							continue;// Not valid to reuse point
+						}
+						real_t d = cp.distance_to(p_pos);
+						if (d < closest_dist && d < grab_threshold) {
+							closest_dist = d;
+							closest_pt = cp;
+							closest_idx = i + 1;
+							found_edge_pt = true;
 						}
 					}
 
-					if (point_drag_index == -1) {
-						uv_drag = false;
+					if (found_edge_pt) {// Add new vertex on edge at selected point
+						uv_create_uv_prev = node->get_uv();
+						uv_create_poly_prev = node->get_polygon();
+						uv_create_prev_internal_vertices = node->get_internal_vertex_count();
+						uv_create_colors_prev = node->get_vertex_colors();
+						uv_create_bones_prev = node->call("_get_bones");
+						polygons_prev = node->get_polygons();
+
+						undo_redo->create_action(TTR("Insert Vertex"));
```

Comment:
Are you sure you need do all of this while only only modifying the polygon ?
Maybe adding a point messing up all of the other properties ?

Reply:
Most of these are needed since they may be changed later, and would be needed when undo/ `_cancel_editing()` is called. I removed the calls to save the bones and polygons since it's not strictly necessary.